Platform/OpenNtfsDxe/Compression.c

```diff
@@ -0,0 +1,668 @@
+  OUT UINT8   *Buffer,
+  IN  UINTN   BlocksTotal
+  )
+{
+  EFI_STATUS   Status;
+  UINTN        SpareBlocks;
+  UINT64       SpareClusters;
+  UINT64       BlocksPerCluster = 0;
+  UINT64       ClustersPerBlock = 0;
+  UINT64       ClearTextClusters;
+
+  BufferSize = BlocksTotal * COMPRESSION_BLOCK;
+
+  if (mClusterSize >= COMPRESSION_BLOCK) {
+    BlocksPerCluster = DivU64x64Remainder (mClusterSize, COMPRESSION_BLOCK, NULL);
+  } else {
+    ClustersPerBlock = DivU64x64Remainder (COMPRESSION_BLOCK, mClusterSize, NULL);
+  }
+
+  while (BlocksTotal != 0) {
+    if ((Runlist->TargetVcn & UNIT_MASK) == 0)  {
+      if ((Runlist->Unit.Head != Runlist->Unit.Tail) && (Runlist->IsSparse == FALSE)) {
+        DEBUG((DEBUG_INFO, "NTFS: Invalid compression block\n"));
+        return EFI_VOLUME_CORRUPTED;
+      }
+
+      Runlist->Unit.Head = Runlist->Unit.Tail = 0;
+      Runlist->Unit.CurrentVcn = Runlist->TargetVcn;
+      Runlist->Unit.ClusterOffset = mClusterSize;
+      if (Runlist->TargetVcn >= Runlist->NextVcn) {
+        Status = ReadRunListElement (Runlist);
+        if (EFI_ERROR (Status)) {
+          return Status;
+        }
+      }
+
+      while (Runlist->TargetVcn + 16 > Runlist->NextVcn) {
+        if (Runlist->IsSparse) {
+          break;
+        }
+
+        Runlist->Unit.Elements[Runlist->Unit.Tail].Vcn = Runlist->NextVcn;
+        Runlist->Unit.Elements[Runlist->Unit.Tail].Lcn = Runlist->CurrentLcn + Runlist->NextVcn - Runlist->CurrentVcn;
+        Runlist->Unit.Tail++;
+
+        Status = ReadRunListElement (Runlist);
+        if (EFI_ERROR (Status)) {
+          return Status;
+        }
+      }
+    }
+
+    if (mClusterSize >= COMPRESSION_BLOCK) {
+      SpareBlocks = (UINTN) ((16 - (Runlist->TargetVcn & UNIT_MASK)) * BlocksPerCluster);
+    } else {
+      SpareBlocks = (UINTN) DivU64x64Remainder (16 - (Runlist->TargetVcn & UNIT_MASK), ClustersPerBlock, NULL);
+    }
+
+    if (SpareBlocks > BlocksTotal) {
+      SpareBlocks = BlocksTotal;
+    }
+    BlocksTotal -= SpareBlocks;
+
+    if (Runlist->IsSparse) {
+      if (mClusterSize >= COMPRESSION_BLOCK) {
+        Runlist->TargetVcn += DivU64x64Remainder (SpareBlocks, BlocksPerCluster, NULL);
+      } else {
+        Runlist->TargetVcn += SpareBlocks * ClustersPerBlock;
+      }
+
+      if (Runlist->Unit.Tail == 0) {
+        if (Buffer != NULL) {
+          if (BufferSize < (SpareBlocks * COMPRESSION_BLOCK)) {
+            DEBUG((DEBUG_INFO, "NTFS: (ReadCompressedBlock #1) Buffer overflow.\n"));
+            return EFI_VOLUME_CORRUPTED;
+          }
+
+          ZeroMem (Buffer, SpareBlocks * COMPRESSION_BLOCK);
+          Buffer += SpareBlocks * COMPRESSION_BLOCK;
+          BufferSize -= SpareBlocks * COMPRESSION_BLOCK;
+        }
+      } else {
+        while (SpareBlocks != 0) {
+          Status = DecompressBlock (&Runlist->Unit, Buffer);
+          if (EFI_ERROR (Status)) {
+            return Status;
+          }
+
+          if (Buffer != NULL) {
```

Comment:
`Buffer` must not be NULL if `DecompressBlock` succeeded.

Reply:
Buffer is OPTIONAL.